[extension/sigv4authextension] Add Unmarshal() to sigv4auth extension (PR)

erichsueh3 opened the pull request:
**Description:** <Describe what has changed.>
This PR addresses an issue brought up in #16490 and refactors the code in `extension/sigv4authextension/config.go`.

**Link to tracking Issue:** #16490

erichsueh3 committed:
add Unmarshal() implementation to sigv4 extension

erichsueh3 committed:
update TestLoadConfigError test

erichsueh3 commented:
Also note that the changelog entry should be skipped as this is code refactoring without any functional change

erichsueh3 committed:
merge conflict resolution

erichsueh3 commented:
Hi @djaglowski anything else I can do to get this PR merged?

erichsueh3 commented:
> Hi @djaglowski anything else I can do to get this PR merged?

Hi @djaglowski, just wanted to follow up on this again

erichsueh3 commented:
> > Hi @djaglowski anything else I can do to get this PR merged?
> 
> Hi @djaglowski, just wanted to follow up on this again

Hi @djaglowski, wanted to check in here again, what can I do to get this merged? It's a small refactor PR that shouldn't cause too much trouble.

atoulme commented on extension/sigv4authextension/config.go:
```diff
@@ -41,10 +42,18 @@ type AssumeRole struct {
 // compile time check that the Config struct satisfies the component.Config interface
 var _ component.Config = (*Config)(nil)
 
-// Validate checks that the configuration is valid.
```
why did you remove that doc?

erichsueh3 replied:
Now that Validate() effectively isn't doing anything, I removed the comment because it wasn't entirely accurate - I can re-add it though, if this is just a consistency thing

atoulme replied:
Validate is still needed looks like, since these errors can be returned:
```
	if err != nil {
		return fmt.Errorf("could not retrieve credential provider: %w", err)
	}
	if credsProvider == nil {
		return fmt.Errorf("credsProvider cannot be nil")
	}
```

My 0.02c: move the checks to a Validate method, and keep this unmarshal method to initialize the config properly.

atoulme commented:
@erichsueh3 please rebase and a changelog entry under .chloggen, copying the TEMPLATE entry.
Please run `make lint` over your changes as well.

erichsueh3 committed:
Merge branch 'main' of https://github.com/erichsueh3/opentelemetry-collector-contrib into sigv4_add_unmarshal

erichsueh3 committed:
changelog

erichsueh3 committed:
use unmarshalconfig() instead of unmarshalextensionconfig()

erichsueh3 commented:
> @erichsueh3 please rebase and a changelog entry under .chloggen, copying the TEMPLATE entry. Please run `make lint` over your changes as well.

Hey @atoulme, I've rebased, added a changelog entry, and ran `make lint` as well, let me know if anything else needs to be done

atoulme commented on .chloggen/sigv4_add_unmarshal.yaml:
```diff
@@ -0,0 +1,16 @@
+component: sigv4authextension
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add Unmarshal() method to config.go
```
Can we get a bit more information in this changelog? If this just an internal change, just remove this changelog and mark the PR as "Skip Changelog"
